app/controllers/identities/omniauth_callbacks_controller.rb

```diff
@@ -20,15 +20,26 @@
 
 class Identities::OmniauthCallbacksController < Devise::OmniauthCallbacksController
   def shibboleth
-    @identity = Identity.find_for_shibboleth_oauth(request.env["omniauth.auth"], current_identity)
+    @identity = Identity.find_for_shibboleth_oauth(request.env['omniauth.auth'], current_identity)
 
     if @identity.persisted?
-      sign_in_and_redirect @identity, :event => :authentication #this will throw if @identity is not activated
-      set_flash_message(:notice, :success, :kind => "Shibboleth") if is_navigational_format?
+      sign_in_and_redirect(@identity, event: :authentication) # this will throw if @identity is not activated
+      set_flash_message(:notice, :success, kind: 'Shibboleth') if is_navigational_format?
     else
-      session["devise.shibboleth_data"] = request.env["omniauth.auth"]
+      session['devise.shibboleth_data'] = request.env['omniauth.auth']
       redirect_to new_identity_registration_url
     end
   end
-end
 
+  def cas
+    @identity = Identity.find_for_cas_oauth(request.env['omniauth.auth'], current_identity)
+
+    if !@identity.nil?
```

Comment:
```ruby
  def self.find_for_cas_oauth(auth, _signed_in_resource = nil)
    Directory.search(auth.uid).first
  end
```

Directory.search will search and then return the first found. If it is not nil, then it is always persisted. Should check nil first. once nil checking passes, no need to check persisted. 

The logic here is that if a user successfully authenticated through CAS, then always succeed. unless this is not always true.